transport/conntable.go

```diff
@@ -36,71 +82,94 @@ func (t *connTable) Notify(addr string, conn *connection) {
-			// if we close unexpectedly.
-			defer func(c *connection) {
-				if c != nil {
-					c.Close()
-				}
-			}(watcher.conn)
-
-			for {
-				select {
-				case <-watcher.timer.C:
-					// Socket expiry timer has run out. Close the connection.
-					log.Debug("Socket %p (%s) inactive for too long; close it", watcher.conn, watcher.addr)
-					watcher.conn.Close()
-					watcher.conn = nil
-				case update := <-watcher.update:
-					// We've been pinged with a connection; update it and refresh the
-					// timer.
-					if update != watcher.conn {
-						log.Debug("Manager for address %s received new socket %p; update records", watcher.addr, watcher.conn)
-						watcher.conn = update
-					}
-					watcher.timer.Reset(c_SOCKET_EXPIRY)
-				case stop := <-watcher.stop:
-					// We've received a termination signal; stop managing this connection.
-					if stop {
-						log.Info("Connection watcher for address %s got the kill signal. Stopping.", watcher.addr)
-						watcher.timer.Stop()
-						watcher.conn.Close()
-						watcher.conn = nil
-						break
-					}
-				}
-			}
-		}(watcher)
+	t.updates <- &connUpdate{addr, conn}
+}
+
+func (t *connTable) handleUpdate(update *connUpdate) {
+	log.Debug("Update received in connTable %p for address %s", t, update.addr)
+	watcher, entry_exists := t.conns[update.addr]
+	if !entry_exists {
+		log.Debug("No connection watcher registered for %s; spawn one", update.addr)
+		watcher = &connWatcher{update.addr, update.conn, timing.NewTimer(c_SOCKET_EXPIRY), make(chan bool), timing.Now().Add(c_SOCKET_EXPIRY), t.expiries, make(chan bool)}
+		t.conns[update.addr] = watcher
+		go watcher.loop()
 	}
 
-	watcher.update <- conn
+	watcher.Update(update.conn)
 }
 
 // Return an existing open socket for the given address, or nil if no such socket
 // exists.
 func (t *connTable) GetConn(addr string) *connection {
-	watcher, ok := t.conns[addr]
-	if ok {
-		log.Debug("Query connection for address %s returns %p", addr, watcher.conn)
-		return watcher.conn
-	} else {
-		log.Debug("Query connection for address %s returns nil (no registered watcher)", addr)
-		return nil
-	}
+	responseChan := make(chan *connection)
+	t.connRequests <- &connRequest{addr, responseChan}
+	conn := <-responseChan
+
+	log.Debug("Query connection for address %s returns %p", conn)
+	return conn
 }
 
 // Close all sockets and stop socket management.
 // The table cannot be restarted after Stop() has been called, and GetConn() will return nil.
 func (t *connTable) Stop() {
-	log.Info("Conntable %p stopped")
-	t.stopped = true
-	for _, watcher := range t.conns {
-		watcher.stop <- true
+	t.stop <- true
+}
+
+// Update the connection associated with a given connWatcher, and reset the
+// timeout timer.
+// Must only be called from the connTable goroutine (and in particular, must
+// *not* be called from the connWatcher goroutine).
+func (watcher *connWatcher) Update(c *connection) {
+	if watcher.timer != nil {
+		watcher.timer.Stop()
```

Comment:
Is it not better here to Reset() the timer instead of creating a whole new one every time?